ipynb/primary-processing/11-generate_per_bodysite_results.md

```diff
@@ -0,0 +1,278 @@
+...     for source, sink in zip(*(source_beta, sink_beta)):
+...         !cp $source $sink
+```
+
+We're going to select a single sample for each subject at each body site. The human microbiome project [doi: ] demonstrated that differences across bodysites are larger than interpersonal differences within a single bodysite. However, other studies (ie Caporaso 2011 [[doi: 10.1186/gb-2011-12-5-r50](http://www.ncbi.nlm.nih.gov/pubmed/21624126)], Wu 2011 [[doi: 10.1126/science](http://www.ncbi.nlm.nih.gov/pubmed/21885731)]) suggest a high degree of stability within individual communities. A single sample can be use to avoid the bias associated with multiple closely related samples from a single individual.
+
+We select the sample using a constrained, random method. Samples which are sequenced at depths greater than the highest rarefaction depth are given priority. However, if multiple samples exist greater than this depth, the sample is selected at random. The process is repeated for each rarefaction depth, until a sample can be identified. This way, the same sample can be retained across analysis performed at multiple depths.
+
+```python
+>>> for bodysite, trim, source_otu, source_map, sink_dir, sink_otu, sink_map in iterate_bodysite():
+...     # We only need to do this once
+...     if not trim == '100nt':
+...         continue
+...
+...     # Reads in the map and adds in the sequencing depth
+...     map_ = pd.read_csv(sink_map, sep='\t', dtype=str).set_index('#SampleID')
+...     depths = !biom summarize-table -i $sink_otu
+...     depths = pd.DataFrame([d.split(': ') for d in depths[15:]], columns=['#SampleID', 'depth']).set_index('#SampleID')
+...     map_ = map_.join(depths)
+...
+...     # Picks a single sample for each individual
+...     single_ids = agu.get_single_id_lists(map_, rare_depths)
+...
+...     # Saves the files
+...     single_fp = agu.get_new_path(agenv.paths['package']['single_ids']['%s-%s' % (bodysite, 'unrare')])
+...     with open(single_fp, 'w') as f_:
+...         f_.write('\n'.join(single_ids['unrare']))
+...
+...     for depth, depth_name in zip(*(rare_depths, depth_names)):
+...         single_fp = agu.get_new_path(agenv.paths['package']['single_ids']['%s-%s' % (bodysite, depth_name)])
+...         with open(single_fp, 'w') as f_:
+...             f_.write('\n'.join(single_ids[depth]))
+```
+
+We're going to define another helper function, to work through the single sample files. We're then going to filter the data.
+
+```python
+>>> def single_sample_iterator():
+...     for bodysite in ['fecal', 'oral', 'skin']:
+...         for depth_name in depth_names:
+...             single_fp = agu.get_existing_path(agenv.paths['package']['single_ids']['%s-%s' % (bodysite, depth_name)])
+...             for trim in ['notrim', '100nt']:
+...                 source_otu = agu.get_existing_path(agenv.paths['package']['all_participants_one_sample'][bodysite][trim]['source-%s-otu' % depth_name])
```

Comment:
Ah.  I see that now.  Cool.

On Fri, Jan 8, 2016 at 11:37 AM, J W Debelius <notifications@github.com>
wrote:

> In ipynb/primary-processing/11-generate_per_bodysite_results.md
> <https://github.com/biocore/American-Gut/pull/182#discussion_r49228385>:
>
> > +...
> > +...     for depth, depth_name in zip(*(rare_depths, depth_names)):
> > +...         single_fp = agu.get_new_path(agenv.paths['package']['single_ids']['%s-%s' % (bodysite, depth_name)])
> > +...         with open(single_fp, 'w') as f_:
> > +...             f_.write('\n'.join(single_ids[depth]))
> > +```
> > +
> > +We're going to define another helper function, to work through the single sample files. We're then going to filter the data.
> > +
> > +```python
> > +>>> def single_sample_iterator():
> > +...     for bodysite in ['fecal', 'oral', 'skin']:
> > +...         for depth_name in depth_names:
> > +...             single_fp = agu.get_existing_path(agenv.paths['package']['single_ids']['%s-%s' % (bodysite, depth_name)])
> > +...             for trim in ['notrim', '100nt']:
> > +...                 source_otu = agu.get_existing_path(agenv.paths['package']['all_participants_one_sample'][bodysite][trim]['source-%s-otu' % depth_name])
>
> This has been addressed. Im not sure why you're seeing the outdated version
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/biocore/American-Gut/pull/182/files#r49228385>.
>